File: backend/core/enums.py
"""
Shared enumeration definitions for backend-frontend synchronization.

These enums ensure consistent values across the entire system and serve as
a single source of truth for signal types, statuses, and states.
"""

from enum import Enum
from typing import List


class SignalType(str, Enum):
    """Trading signal type enumeration.
    
    Canonical labels use LONG/SHORT. Legacy BUY/SELL aliases are accepted via
    ``normalize()`` and ``is_valid()`` for backward compatibility.
    
    Must match frontend SignalType in frontend/types/enums.ts
    """
    STRONG_LONG = "STRONG_LONG"
    LONG = "LONG"
    HOLD = "HOLD"
    SHORT = "SHORT"
    STRONG_SHORT = "STRONG_SHORT"
    
    @classmethod
    def get_valid_values(cls) -> List[str]:
        """Get list of valid signal values."""
        return [member.value for member in cls]
    
    @classmethod
    def is_valid(cls, value: str) -> bool:
        """Check if value is a valid signal (canonical or legacy alias)."""
        from agent.core.signal_vocabulary import normalize_signal
        normalized = normalize_signal(value, default="")
        return normalized in cls.get_valid_values()
    
    @classmethod
    def normalize(cls, value: str, default: str = "HOLD") -> str:
        """Normalize and validate a signal value.
        
        Args:
            value: Signal value to normalize
            default: Default value if invalid
            
        Returns:
            Valid canonical signal value
        """
        from agent.core.signal_vocabulary import normalize_signal
        normalized = normalize_signal(value, default=default)
        if normalized in cls.get_valid_values():
            return normalized
        return default


class PositionStatus(str, Enum):
    """Position status enumeration.
    
    Must match backend PositionStatus in backend/core/database.py
    """
    OPEN = "OPEN"
    CLOSED = "CLOSED"
    LIQUIDATED = "LIQUIDATED"


class TradeStatus(str, Enum):
    """Trade status enumeration.
    
    Must match backend TradeStatus in backend/core/database.py
    """
    EXECUTED = "EXECUTED"
    PENDING = "PENDING"
    FAILED = "FAILED"


class TradeSide(str, Enum):
    """Trade side enumeration (BUY or SELL).
    
    Must match backend TradeSide in backend/core/database.py
    """
    BUY = "BUY"
    SELL = "SELL"


class PositionSide(str, Enum):
    """Position side enumeration (LONG or SHORT)."""
    LONG = "LONG"
    SHORT = "SHORT"


class AgentState(str, Enum):
    """Agent state enumeration."""
    UNKNOWN = "UNKNOWN"
    INITIALIZING = "INITIALIZING"
    MONITORING = "MONITORING"
    OBSERVING = "OBSERVING"
    DECISION_MAKING = "DECISION_MAKING"
    EXECUTING = "EXECUTING"
    PAUSED = "PAUSED"
    STOPPED = "STOPPED"
    ERROR = "ERROR"


class HealthStatus(str, Enum):
    """System health status."""
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


class ServiceStatus(str, Enum):
    """Individual service status."""
    UP = "up"
    DOWN = "down"
    DEGRADED = "degraded"


# Simplified WebSocket message types
class WebSocketMessageType(str, Enum):
    """Consolidated WebSocket message types."""
    DATA_UPDATE = "data_update"  # Replaces signal_update, portfolio_update, trade_executed, etc.
    AGENT_UPDATE = "agent_update"  # Replaces agent_state
    SYSTEM_UPDATE = "system_update"  # Replaces health_update, time_sync
    RESPONSE = "response"  # For request-response pattern
    ERROR = "error"  # Error messages


class WebSocketResource(str, Enum):
    """WebSocket resource types for data_update messages."""
    SIGNAL = "signal"  # Trading signals and reasoning
    PORTFOLIO = "portfolio"  # Portfolio and positions
    TRADE = "trade"  # Trade executions
    MARKET = "market"  # Market data and ticks
    PERFORMANCE = "performance"  # Performance metrics
    HEALTH = "health"  # System health
    TIME = "time"  # Time synchronization
    AGENT = "agent"  # Agent state
    MODEL = "model"  # Model predictions


# Confidence thresholds for signal mapping
SIGNAL_CONFIDENCE_THRESHOLDS = {
    "STRONG_LONG_MIN": 0.8,
    "LONG_MIN": 0.65,
    "SHORT_MAX": 0.35,
    "STRONG_SHORT_MAX": 0.2,
}


def get_signal_from_confidence(confidence: float) -> str:
    """Map confidence value to signal type.
    
    Args:
        confidence: Confidence value (0.0 to 1.0)
        
    Returns:
        Corresponding SignalType value
    """
    if confidence >= SIGNAL_CONFIDENCE_THRESHOLDS["STRONG_LONG_MIN"]:
        return SignalType.STRONG_LONG.value
    elif confidence >= SIGNAL_CONFIDENCE_THRESHOLDS["LONG_MIN"]:
        return SignalType.LONG.value
    elif confidence <= SIGNAL_CONFIDENCE_THRESHOLDS["STRONG_SHORT_MAX"]:
        return SignalType.STRONG_SHORT.value
    elif confidence <= SIGNAL_CONFIDENCE_THRESHOLDS["SHORT_MAX"]:
        return SignalType.SHORT.value
    else:
        return SignalType.HOLD.value

File: frontend/types/enums.ts
/**
 * Shared enumerations for backend-frontend type synchronization.
 * 
 * These enums are derived from backend definitions and ensure that both
 * frontend and backend use the same values for signals, statuses, and states.
 */

/**
 * Trading signal types - must match backend signal definitions.
 * 
 * @see backend/services/agent_event_subscriber.py (valid_signals list)
 * @see backend/api/routes/trading.py (signal mapping logic)
 */
export const SIGNAL_TYPE_VALUES = ['STRONG_BUY', 'BUY', 'HOLD', 'SELL', 'STRONG_SELL'] as const;
export type SignalType = typeof SIGNAL_TYPE_VALUES[number];

export const isValidSignal = (value: unknown): value is SignalType => {
  return SIGNAL_TYPE_VALUES.includes(value as SignalType);
};

/**
 * Position status - must match backend PositionStatus enum.
 * 
 * @see backend/core/database.py (PositionStatus enum)
 */
export const POSITION_STATUS_VALUES = ['OPEN', 'CLOSED', 'LIQUIDATED'] as const;
export type PositionStatus = typeof POSITION_STATUS_VALUES[number];

export const isValidPositionStatus = (value: unknown): value is PositionStatus => {
  return POSITION_STATUS_VALUES.includes(value as PositionStatus);
};

/**
 * Trade status - must match backend TradeStatus enum.
 * 
 * @see backend/core/database.py (TradeStatus enum)
 */
export const TRADE_STATUS_VALUES = ['EXECUTED', 'PENDING', 'FAILED'] as const;
export type TradeStatus = typeof TRADE_STATUS_VALUES[number];

export const isValidTradeStatus = (value: unknown): value is TradeStatus => {
  return TRADE_STATUS_VALUES.includes(value as TradeStatus);
};

/**
 * Trade side - must match backend TradeSide enum.
 * 
 * @see backend/core/database.py (TradeSide enum)
 */
export const TRADE_SIDE_VALUES = ['BUY', 'SELL'] as const;
export type TradeSide = typeof TRADE_SIDE_VALUES[number];

export const isValidTradeSide = (value: unknown): value is TradeSide => {
  return TRADE_SIDE_VALUES.includes(value as TradeSide);
};

/**
 * Position side - can be LONG or SHORT.
 */
export const POSITION_SIDE_VALUES = ['LONG', 'SHORT'] as const;
export type PositionSide = typeof POSITION_SIDE_VALUES[number];

export const isValidPositionSide = (value: unknown): value is PositionSide => {
  return POSITION_SIDE_VALUES.includes(value as PositionSide);
};

/**
 * Agent state values - various states the agent can be in.
 */
export const AGENT_STATE_VALUES = [
  'UNKNOWN',
  'INITIALIZING',
  'MONITORING',
  'OBSERVING',
  'DECISION_MAKING',
  'EXECUTING',
  'PAUSED',
  'STOPPED',
  'ERROR',
] as const;
export type AgentState = typeof AGENT_STATE_VALUES[number];

export const isValidAgentState = (value: unknown): value is AgentState => {
  return AGENT_STATE_VALUES.includes(value as AgentState);
};

/**
 * Health status values.
 */
export const HEALTH_STATUS_VALUES = ['healthy', 'degraded', 'unhealthy'] as const;
export type HealthStatusType = typeof HEALTH_STATUS_VALUES[number];

export const isValidHealthStatus = (value: unknown): value is HealthStatusType => {
  return HEALTH_STATUS_VALUES.includes(value as HealthStatusType);
};

/**
 * Service status values.
 */
export const SERVICE_STATUS_VALUES = ['up', 'down', 'degraded', 'unknown'] as const;
export type ServiceStatusType = typeof SERVICE_STATUS_VALUES[number];

export const isValidServiceStatus = (value: unknown): value is ServiceStatusType => {
  return SERVICE_STATUS_VALUES.includes(value as ServiceStatusType);
};

/**
 * Validate and normalize a signal type.
 *
 * @param signal Raw signal value from backend
 * @returns Valid SignalType or null if invalid
 */
export function normalizeSignalType(signal: unknown): SignalType | null {
  if (isValidSignal(signal)) {
    return signal;
  }
  console.warn(`Invalid signal type: ${signal}, treating as no signal`);
  return null;
}

/**
 * Get human-readable label for a signal type.
 */
export function getSignalLabel(signal: SignalType): string {
  const labels: Record<SignalType, string> = {
    'STRONG_BUY': 'Strong Buy',
    'BUY': 'Buy',
    'HOLD': 'Hold',
    'SELL': 'Sell',
    'STRONG_SELL': 'Strong Sell',
  };
  return labels[signal] || 'Unknown';
}

/**
 * Get color for a signal type (for UI rendering).
 */
export function getSignalColor(signal: SignalType): string {
  const colors: Record<SignalType, string> = {
    'STRONG_BUY': '#00c853',  // Green
    'BUY': '#66bb6a',         // Light Green
    'HOLD': '#ffa726',        // Orange
    'SELL': '#ef5350',        // Light Red
    'STRONG_SELL': '#c62828', // Dark Red
  };
  return colors[signal] || '#9e9e9e';
}

/**
 * Get numeric confidence threshold for signal strength.
 * Used for interpreting raw predictions to signals.
 */
export const SIGNAL_CONFIDENCE_THRESHOLDS = {
  STRONG_BUY_MIN: 0.8,
  BUY_MIN: 0.65,
  SELL_MAX: 0.35,
  STRONG_SELL_MAX: 0.2,
} as const;
